### DISCUSSION TITLE
pandas category to ENUM

### DISCUSSION THREAD

**Discussion by ffyring** [duckdb/duckdb#5738] [General]
Currently, it does not seem like Pandas categorical values survive a roundtrip if we save the table to parquet. E.g.

```
# Our unencoded data.
data = ['Hobbit', 'Elf', 'Elf', 'Man', 'Mayar', 'Hobbit', 'Mayar']
# 'pd.Categorical' automatically encodes the data as a categorical column
df_in = pd.DataFrame({'races': pd.Categorical(data),})
# We can query this dataframe as we would any other
# The conversion from categorical columns to enums happens automatically
df_out = duckdb.execute("SELECT * FROM df_in").df()

>> df_out.info()
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   races   7 non-null      category
```

Very good this far - the categories survive getting them out of DuckDB. Now we save as parquet:
```
uckdb.execute("COPY df_in to 'df_category.parquet' (FORMAT PARQUET)")
df_out2 = duckdb.execute("SELECT * from 'df_category.parquet'").df()

>> df_out2.info()
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   races   7 non-null      object
```

So unfortunately the categories don't survive this way. I tried to export database like
```
# Export whole database
duckdb.execute('CREATE OR REPLACE TABLE df2 AS SELECT * FROM df_in')
duckdb.execute("EXPORT DATABASE 'duckdb_db';")
# Import database to check if ENUM survives
duckdb.execute("IMPORT DATABASE 'duckdb_db';")

>>
InvalidInputException: Invalid Input Error: Attempting to execute an unsuccessful or closed pending query result
Error: Catalog Error: Type with name races does not exist!
Did you mean "date"?
```

So no luck there. 
Is it possible, or planned, to be able to save a table with categorical values to parquet? Parquet itself supports it (see e.g. https://stackoverflow.com/questions/54731396/pandas-dataframe-with-categorical-columns-from-a-parquet-file-using-read-parquet)

Kind Regards
 Fredrik

**Comment by Mytherin** [comment]
Parquet does not support categorical values. However, the Arrow Parquet library has "sort of" added support for this into Parquet files by storing their own custom schema inside the generic metadata field that Parquet files have. See #3309. 

While that works - it is not portable across libraries. The categorical type is only consistent when read and written by the same library, i.e. when reading a Parquet file produced by Arrow in a non-Arrow reader the categorical type is lost again.

We could do something similar in our reader/writer which would allow DuckDB to round-trip enum values to Parquet - but that would then be DuckDB specific, i.e. files produced by DuckDB would maintain the categorical status only when then read by DuckDB again.
